### DISCUSSION TITLE
Reimplementing DualNet using Avalanche

### DISCUSSION THREAD

**Discussion by zafstojano** [ContinualAI/avalanche#909] [Q&A]
Dear Avalanche community,

I hope you are doing great. 

This might be a newbie question, but I need your help. I would like to implement DualNet(https://arxiv.org/abs/2110.00175) using Avalanche, and I am trying to first think through the necessary components that I would need.

Given the open-source implementation of the method (https://github.com/phquang/DualNet), I believe I need to achieve the following:
1. Implement a replay memory, that stores inputs x, targets y, and features t (extracted from the inputs x) at the end of each task.
2. Due to the computation of the loss function, I would need to split the stream of current task examples, and previous task examples. To achieve this, I found this discussion (https://github.com/ContinualAI/avalanche/discussions/644) very helpful. 

My question is related to step 1. 

What is the best way to implement a Reservoir to store the triplets (x,y,t), given the requirements of step 2 (separability)?

**Comment by vlomonaco** [comment]
Hi @zaf-stojano! Thanks for your question. We have an in-depth tutorial on [how to handle replay in Avalanche](https://avalanche.continualai.org/how-tos/dataloading_buffers_replay). There are a few data loaders already implemented that may help you in what you are trying to achieve. Let us know if something is still unclear! :)

**Reply by zafstojano** [reply]
Thanks, I'll have a look :)
